### PR TITLE
Fix multimodal cache hash collisions causing embedding corruption and DoS

### PR DESCRIPTION
**Pull request by 3em0**
## Summary

Fixes multiple hash collision vulnerabilities in the multimodal embedding cache system that cause deterministic embedding corruption, DoS crashes, and silent cache invalidation in multi-process deployments.

Closes #22032

### 1. `data_hash`: Extend hash width from 64-bit to 128-bit
- `hashlib.sha256(data).digest()[:8]` → `[:16]`
- Raises birthday-paradox collision threshold from ~2^32 to ~2^64

### 2. `tensor_hash`: Add shape/dtype boundary markers
- Before hashing each tensor's bytes, prepend `"{shape}:{dtype}|"` as a separator
- Prevents `[[1,2],[3,4,5]]` and `[[1,2,3],[4,5]]` from producing identical hashes

### 3. `hash_feature` (ndarray branch): Include shape and dtype metadata
- Prepend `"{shape}:{dtype}:"` before raw array bytes
- Prevents arrays like `shape=(16,)` and `shape=(4,4)` with same values from colliding

### 4. `combine_hashes`: Replace Python `hash()` with deterministic SHA256
- Python's built-in `hash()` is randomized per process (PYTHONHASHSEED)
- Now uses `hashlib.sha256` for consistent results across TP workers, restarts, and distributed nodes

### 5. `hash_feature` (list branch): Fix TypeError on non-tensor lists
- `data_hash(tuple(...))` crashes because `tuple` is not bytes-like
- Now properly serializes `int`/`float`/`str` elements via `struct.pack`
- Also handles empty lists gracefully

### 6. `precomputed_embedding`: Add input validation
- Validates that feature is a tensor (converts list/ndarray automatically)
- Rejects non-tensor types and <2D tensors

## Test plan

- [ ] Verify existing multimodal tests pass (no hash regression)
- [ ] Test multi-image inference produces correct per-image embeddings
- [ ] Test ndarray features with different shapes hash differently
- [ ] Test `hash_feature([1, 2, 3])` no longer crashes
- [ ] Test `combine_hashes` produces same result across process restarts

---
### CI States

Latest PR Test (Base): :x: [Run #26558232426](https://github.com/sgl-project/sglang/actions/runs/26558232426)
Latest PR Test (Extra): :x: [Run #26558232382](https://github.com/sgl-project/sglang/actions/runs/26558232382)

### CODE REVIEW

**Comment by abinggo** [comment]
Read through the diff — this looks like a well-structured fix that covers all 6 issues re-raised in #25462. Notes:

**LGTM aspects**
- Promoting the truncated SHA256 from 64 bits to 128 bits closes the birthday-collision exposure cleanly.
- `_mix_shape_with_gpu_hash` is the right shape — the GPU triton path is byte-bucket-blind, so layering a CPU-side shape/dtype mix on top of it preserves the fast kernel while killing the boundary collision class.
- Switching `combine_hashes` from `hash()` to deterministic SHA256 fixes the cross-process `PYTHONHASHSEED` instability — which is the most user-visible issue (cache hits randomly missing across server restarts).
- `precomputed_embedding` input validation is the right place to gate user-controlled cache keys.

**Two minor edge cases worth a follow-up commit**

1. `tensor_hash` GPU branch passes `tensor_list` (not the cleaned `tensors` list) into `_mix_shape_with_gpu_hash`:
   ```python
   tensors = [x.flatten() if isinstance(x, torch.Tensor) else x for x in tensor_list]
   ...
   raw_hash = gpu_tensor_hash(concat_tensor.cuda())
   return _mix_shape_with_gpu_hash(tensor_list, raw_hash)
   ```
   If any element of `tensor_list` is non-tensor, `_mix_shape_with_gpu_hash` hits `AttributeError` on `.shape`/`.dtype`. Either pass `tensors` (with type guards inside `_mix_shape_with_gpu_hash`) or assert the GPU-branch input is all-tensor at entry.

2. `hash_feature`'s nested `serialize`:
   - `bool` is a subclass of `int`, so `isinstance(True, int)` is True — `hash_feature([True])` collides with `hash_feature([1])`. Add an explicit `isinstance(item, bool)` check before the int branch.
   - No `None` handling — `str(None)` falls into the catch-all string branch as `"None"`, colliding with the literal `"None"`. A sentinel `b"n"` for None is one extra line.

   Low-frequency in practice but they harden the boundary for unusual mm payloads.

**Cache compatibility note** (worth flagging in PR body or release notes)

Going from 64-bit to 128-bit hash widths means all currently-stored multimodal cache keys in deployed instances will miss after this lands. Effect is one-time cache cold-start, not data loss, but worth surfacing for ops running long-lived servers.

Independent of these, the direction is right. The two issue threads (#8342 from July and now #25462) make a strong case for urgency. Hopefully a maintainer can `run-ci`-label this so it can move forward.

**Comment by 3em0** [comment]
@yhyang201 @mickqian

**Comment by adarshxs** [comment]
@3em0 can you resolve conflicts?

**Comment by adarshxs** [comment]
@3em0 thanks for raising the PR. do you have any performance benchmarks for the changes?

**Comment by 3em0** [comment]
> @3em0 thanks for raising the PR. do you have any performance benchmarks for the changes?

ok,i will resolve conflicts. But I don’t understand whether the “benchmark” you’re referring to here means the impact on program performance, or how to measure whether this PR has completed the vulnerability fix.

**Comment by adarshxs** [comment]
@3em0 yes I meant impact on performance in terms of speed - individual unit tests for the affected paths and overall. It would be nice to have some hash performance results example: https://github.com/sgl-project/sglang/pull/5974#issue-3035929332

**Comment by 3em0** [comment]
> @3em0 yes I meant impact on performance in terms of speed - individual unit tests for the affected paths and overall. It would be nice to have some hash performance results example: [#5974 (comment)](https://github.com/sgl-project/sglang/pull/5974#issue-3035929332)

ok! i know ! i will do it!

**Comment by 3em0** [comment]
@adarshxs Thanks for the pointer to #5974 — here are the hash performance numbers in the same style. All measurements are CPU-only; the only GPU-side change in this PR is the `_mix_shape_with_gpu_hash` wrapper which reads `tensor.shape` / `tensor.dtype` on CPU (covered by the tensor cases below).

### Hash performance (best of 3 wall-clock, 1000 calls each)

| scenario | old | new | Δ |
|---|---|---|---|
| `tensor_hash` CPU, sgl-logo image `(3312, 1176)` float32 | 23.708 s | 23.526 s | -0.8% |
| `tensor_hash` CPU, tensor `(196, 768)` float32 | 916.65 ms | 918.39 ms | +0.2% |
| `tensor_hash` CPU, tensor `(576, 1024)` float32 | 3.584 s | 3.595 s | +0.3% |
| `tensor_hash` CPU, tensor `(1024, 1024)` float32 | 6.366 s | 6.402 s | +0.6% |
| `hash_feature` ndarray `(196, 768)` float32 | 914.43 ms | 921.35 ms | +0.8% |
| `hash_feature` ndarray `(576, 1024)` float32 | 3.594 s | 3.596 s | +0.0% |
| `hash_feature` `list[2 × tensor (196, 768)]` | 1.845 s | 1.849 s | +0.2% |
| `hash_feature` `list[4 × tensor (196, 768)]` | 3.689 s | 3.691 s | +0.1% |
| `combine_hashes` n=1 | 52.4 µs | 605.0 µs | +1053.6% |
| `combine_hashes` n=4 | 59.2 µs | 903.0 µs | +1424.6% |
| `combine_hashes` n=16 | 93.5 µs | 1.83 ms | +1855.0% |
| `data_hash` 64-byte payload | 512.7 µs | 515.3 µs | +0.5% |
| `data_hash` 1024-byte payload | 1.97 ms | 1.96 ms | -0.2% |

### Correctness regression fixed by this PR

`hash_feature([1, 2, ..., 63])` (list of 64 ints):

- pre-PR: **TypeError** — `object supporting the buffer API required`
- post-PR: ok, 8.88 ms for 1000 calls

### Summary

- **Multimodal payload hash paths (`tensor_hash`, `hash_feature` for ndarray/tensor/list-of-tensors) are within ±1%** of pre-PR. The added shape/dtype boundary prefix is a few tens of bytes vs MB-scale payloads, so SHA256 throughput dominates and the overhead is unobservable.
- **`combine_hashes` shows a large percentage delta but the absolute is microseconds.** It's called *once* per multimodal request (request processing is in the ms+ range), so the cost is invisible end-to-end. The change is required to fix the `PYTHONHASHSEED` cross-process inconsistency that caused silent cache misses across TP workers / restarts (the user-visible issue from #25462).
- **List-of-int input**: pre-PR raised `TypeError` (fix #5 in PR body); post-PR works correctly.

<details>
<summary><b>Reproduction script (click to expand)</b> — save as <code>bench_hash_collision_fix.py</code>, run with <code>python bench_hash_collision_fix.py</code></summary>

```python
"""
CPU-only micro-benchmark for PR #22033 (multimodal cache hash collision fix).

Compares the pre-PR ("old") and post-PR ("new") implementations of every
hash function the PR touches:

  - data_hash        SHA256 truncated 8 -> 16 bytes
  - combine_hashes   Python hash() -> deterministic SHA256
  - tensor_hash      CPU path adds shape/dtype boundary marker
  - hash_feature     ndarray gains shape/dtype prefix; list path uses
                     struct.pack with type/structure tags

GPU paths are out of scope: this PR's only GPU-side addition is the CPU
wrapper `_mix_shape_with_gpu_hash`, which reads `t.shape` / `t.dtype`
only and is exercised by the tensor cases below.

Run:  python bench_hash_collision_fix.py
"""

import hashlib
import struct
import time

import numpy as np
import torch


def flatten_nested_list(lst):
    out = []
    for x in lst:
        if isinstance(x, list):
            out.extend(flatten_nested_list(x))
        else:
            out.append(x)
    return out


# ---------- OLD (pre-PR) implementations ----------

def data_hash_old(data):
    return int.from_bytes(hashlib.sha256(data).digest()[:8], "big", signed=False)


def combine_hashes_old(mm_hashes):
    if not mm_hashes:
        return None
    return hash(tuple(mm_hashes))


def tensor_hash_old_cpu(tensor_list):
    if isinstance(tensor_list, list):
        tensor_list = flatten_nested_list(tensor_list)
        tensors = [x.flatten() for x in tensor_list]
        hasher = hashlib.sha256()
        for t in tensors:
            t = t.detach().contiguous()
            hasher.update(memoryview(t.view(torch.uint8).numpy()))
        return int.from_bytes(hasher.digest()[:8], "big", signed=False)
    tensor = tensor_list.detach().contiguous()
    hasher = hashlib.sha256()
    hasher.update(memoryview(tensor.view(torch.uint8).numpy()))
    return int.from_bytes(hasher.digest()[:8], "big", signed=False)


def hash_feature_old(f):
    if isinstance(f, list):
        if isinstance(f[0], torch.Tensor):
            return tensor_hash_old_cpu(f)
        # Buggy pre-PR path — crashes for list[int|float|str].
        return data_hash_old(tuple(flatten_nested_list(f)))
    if isinstance(f, np.ndarray):
        arr = np.ascontiguousarray(f)
        hasher = hashlib.sha256()
        hasher.update(memoryview(arr))
        return int.from_bytes(hasher.digest()[:8], "big", signed=False)
    if isinstance(f, torch.Tensor):
        return tensor_hash_old_cpu([f])
    return data_hash_old(f)


# ---------- NEW (post-PR) implementations ----------

def data_hash_new(data):
    return int.from_bytes(hashlib.sha256(data).digest()[:16], "big", signed=False)


def combine_hashes_new(mm_hashes):
    if not mm_hashes:
        return None
    data = b"".join(h.to_bytes(16, "big", signed=False) for h in mm_hashes)
    return int.from_bytes(hashlib.sha256(data).digest()[:16], "big")


def tensor_hash_new_cpu(tensor_list):
    if isinstance(tensor_list, list):
        tensor_list = flatten_nested_list(tensor_list)
        tensors = [x.flatten() for x in tensor_list]
        hasher = hashlib.sha256()
        for t in tensors:
            t = t.detach().contiguous()
            tag = f"{tuple(t.shape)}:{t.dtype}|".encode()
            hasher.update(tag)
            hasher.update(memoryview(t.view(torch.uint8).numpy()))
        return int.from_bytes(hasher.digest()[:16], "big", signed=False)
    tensor = tensor_list.detach().contiguous()
    hasher = hashlib.sha256()
    tag = f"{tuple(tensor.shape)}:{tensor.dtype}|".encode()
    hasher.update(tag)
    hasher.update(memoryview(tensor.view(torch.uint8).numpy()))
    return int.from_bytes(hasher.digest()[:16], "big", signed=False)


def hash_feature_new(f):
    if isinstance(f, list):
        if len(f) == 0:
            return data_hash_new(b"empty_list")
        if isinstance(f[0], torch.Tensor):
            return tensor_hash_new_cpu(f)
        parts = []

        def serialize(item):
            if isinstance(item, list):
                parts.append(b"[")
                for i in item:
                    serialize(i)
                parts.append(b"]")
            elif isinstance(item, float):
                parts.append(b"f" + struct.pack(">d", item))
            elif isinstance(item, int):
                parts.append(b"i" + struct.pack(">q", item))
            else:
                encoded = str(item).encode()
                parts.append(b"s" + len(encoded).to_bytes(4, "big") + encoded)

        serialize(f)
        return data_hash_new(b"".join(parts))
    if isinstance(f, np.ndarray):
        arr = np.ascontiguousarray(f)
        meta = f"{arr.shape}:{arr.dtype}:".encode()
        hasher = hashlib.sha256()
        hasher.update(meta)
        hasher.update(memoryview(arr))
        return int.from_bytes(hasher.digest()[:16], "big", signed=False)
    if isinstance(f, torch.Tensor):
        return tensor_hash_new_cpu([f])
    return data_hash_new(f)


# ---------- bench harness ----------

N_ITERS = 1000  # matches PR #5974 reporting style


def bench_total(fn, args, n_iters=N_ITERS, repeat=3):
    for _ in range(3):
        fn(*args)
    times = []
    for _ in range(repeat):
        t0 = time.perf_counter()
        for _ in range(n_iters):
            fn(*args)
        times.append(time.perf_counter() - t0)
    return min(times)


def fmt_total(s):
    if s < 1e-3:
        return f"{s * 1e6:.1f} us"
    if s < 1:
        return f"{s * 1e3:.2f} ms"
    return f"{s:.3f} s"


def main():
    rng = np.random.RandomState(0)
    torch.manual_seed(0)

    # PR #5974 used a sgl-logo-sized image feature: shape=[3312, 1176] float32.
    sgl_logo_shape = (3312, 1176)

    scenarios = []

    # --- tensor_hash CPU path (parallel to #5974's hash-perf measurement) ---
    sgl_logo_tensor = torch.randn(*sgl_logo_shape, dtype=torch.float32)
    scenarios.append((
        f"tensor_hash CPU, sgl-logo image {sgl_logo_shape} float32",
        lambda: tensor_hash_old_cpu(sgl_logo_tensor),
        lambda: tensor_hash_new_cpu(sgl_logo_tensor),
    ))
    for shape in ((196, 768), (576, 1024), (1024, 1024)):
        t = torch.randn(*shape, dtype=torch.float32)
        scenarios.append((
            f"tensor_hash CPU, tensor {shape} float32",
            lambda t=t: tensor_hash_old_cpu(t),
            lambda t=t: tensor_hash_new_cpu(t),
        ))

    # --- hash_feature (ndarray path) ---
    for shape in ((196, 768), (576, 1024)):
        arr = rng.rand(*shape).astype(np.float32)
        scenarios.append((
            f"hash_feature ndarray {shape} float32",
            lambda arr=arr: hash_feature_old(arr),
            lambda arr=arr: hash_feature_new(arr),
        ))

    # --- hash_feature (multi-image list path) ---
    for n_imgs, shape in [(2, (196, 768)), (4, (196, 768))]:
        ts = [torch.randn(*shape, dtype=torch.float32) for _ in range(n_imgs)]
        scenarios.append((
            f"hash_feature list[{n_imgs} x tensor {shape}]",
            lambda ts=ts: hash_feature_old(ts),
            lambda ts=ts: hash_feature_new(ts),
        ))

    # --- combine_hashes (hot path, called once per request) ---
    for n_items in (1, 4, 16):
        hashes = [2**63 + i for i in range(n_items)]
        scenarios.append((
            f"combine_hashes n_items={n_items}",
            lambda hashes=hashes: combine_hashes_old(hashes),
            lambda hashes=hashes: combine_hashes_new(hashes),
        ))

    # --- data_hash on small payloads ---
    for n in (64, 1024):
        data = b"x" * n
        scenarios.append((
            f"data_hash {n}-byte payload",
            lambda data=data: data_hash_old(data),
            lambda data=data: data_hash_new(data),
        ))

    print(f"### Hash performance (best of 3 wall-clock, {N_ITERS} calls each)\n")
    print("| scenario | old | new | Δ |")
    print("|---|---|---|---|")
    for label, old_fn, new_fn in scenarios:
        old_t = bench_total(old_fn, ())
        new_t = bench_total(new_fn, ())
        delta = (new_t - old_t) / old_t * 100
        print(
            f"| {label} | {fmt_total(old_t)} | {fmt_total(new_t)} | "
            f"{delta:+.1f}% |"
        )

    print("\n### Correctness regression fixed by this PR\n")
    list_input = list(range(64))
    print("`hash_feature([1, 2, ..., 63])` (list of 64 ints):\n")
    try:
        hash_feature_old(list_input)
        print("- pre-PR: ok  (unexpected)")
    except TypeError as e:
        print(f"- pre-PR: **TypeError** — `{e}`")
    new_t = bench_total(hash_feature_new, (list_input,))
    print(f"- post-PR: ok, {fmt_total(new_t)} for {N_ITERS} calls")


if __name__ == "__main__":
    main()
```

</details>

**Comment by 3em0** [comment]
@adarshxs review when you get a chance — bug is fixed. Heads up on `combine_hashes`: feel free to revert that change if you'd rather optimize for performance.
